File: frontend/src/features/reader/reading-clock.ts
"use client";

import { useCallback, useEffect, useRef } from "react";

/**
 * How long the reader has been reading, as the progress endpoint wants it.
 *
 * `POST /reader/progress` takes `time_spent_seconds` as the wall-clock time
 * spent SINCE THIS CLIENT'S LAST PUSH, not a running total — the server
 * accumulates it onto the chapter's row and turns each advance's delta into a
 * `ReadingSession`'s duration (`progress_service.py`). A cumulative figure
 * would be re-added on every push.
 *
 * No client was sending it at all, so every session was recorded with
 * `started_at == ended_at` and every duration derived from them was zero. The
 * statistics screen's "Time read" was therefore not merely wrong, it was
 * structurally incapable of being anything but 0 — while pages and streaks
 * carried on working, so the screen looked alive.
 */

/**
 * The longest gap between two pushes that is still counted as reading.
 *
 * The reader pushes on a 500 ms debounce whenever the position advances, so a
 * gap of minutes means nobody was reading — a locked screen, another tab, a
 * chapter left open overnight. Counting the cap rather than the gap keeps a
 * genuinely slow page honest and refuses to bill an idle night as reading. It
 * errs low on purpose: an under-count is a smaller lie than nine hours of
 * "reading" a locked phone.
 */
export const MAX_PUSH_SECONDS = 300;

/** Whole seconds between two instants, floored at 0 and capped at `cap`. */
export function elapsedSince(
  lastAt: number | null,
  now: number,
  cap: number = MAX_PUSH_SECONDS,
): number {
  if (lastAt === null || !Number.isFinite(lastAt) || !Number.isFinite(now)) return 0;
  // A clock that jumped backwards (NTP correction, a device waking) reports
  // nothing rather than a negative the server would clamp away anyway.
  const seconds = Math.floor((now - lastAt) / 1000);
  if (seconds <= 0) return 0;
  return Math.min(seconds, cap);
}

/**
 * `take()` — the seconds read since the last time it was asked, and reset.
 *
 * One clock per reader rather than one per chapter: the time between two
 * pushes was spent reading whatever the push being made is about, including
 * the chapter just finished at a seam. Attributing it any more finely would
 * need to know when the reading line crossed, which is exactly the number the
 * push is reporting.
 */
export function useReadingClock(): () => number {
  const lastAt = useRef<number | null>(null);

  // Started at mount, so the first push carries the time spent on the first
  // page instead of zero.
  useEffect(() => {
    lastAt.current = Date.now();
  }, []);

  return useCallback(() => {
    const now = Date.now();
    const seconds = elapsedSince(lastAt.current, now);
    lastAt.current = now;
    return seconds;
  }, []);
}

File: mobile/lib/features/reader/utils/reading_clock.dart
/// Longest gap between two progress pushes that still counts as reading.
///
/// A reader left open is not a reader being read: the phone locks, the app is
/// backgrounded, or the book sits on the table until morning. Capping
/// under-reports one genuinely slow page and refuses to invent an afternoon,
/// which is the right way round for a statistic.
const int kMaxReadingGapSeconds = 300;

/// Wall-clock reading time between two progress pushes.
///
/// The server accumulates `time_spent_seconds` and derives every reading
/// session's duration from it (`_apply_one` in
/// `backend/services/progress_service.py` offsets `started_at` by the delta),
/// so a client that never sends one leaves the statistics screen's time-read
/// figures structurally zero rather than merely wrong. This is the client's
/// half of it.
///
/// [elapsed] returns the delta since the previous call, never a running
/// total: the server adds what it is sent, so a cumulative figure would
/// inflate on every replayed push.
class ReadingClock {
  ReadingClock(DateTime startedAt) : _mark = startedAt;

  DateTime _mark;

  /// Seconds read since the previous call, or since the reader opened.
  int elapsed(DateTime now) {
    final gap = now.difference(_mark);
    if (gap.isNegative) {
      // The device clock moved backwards (NTP, a timezone-less manual set).
      // Crediting a negative reading session is worse than crediting none.
      _mark = now;
      return 0;
    }
    if (gap.inSeconds > kMaxReadingGapSeconds) {
      _mark = now;
      return kMaxReadingGapSeconds;
    }
    // Advance the mark only by the whole seconds actually credited, so the
    // sub-second remainder carries into the next push. Pages settle every
    // 500ms; truncating each time would report zero for a fast read.
    final seconds = gap.inSeconds;
    _mark = _mark.add(Duration(seconds: seconds));
    return seconds;
  }
}
